### PR TITLE
Adding the following models: Vicuna, Koala, Pythia-Chat-Base-7B, GPT-NeoXT-Chat-Base-20B 

### PR DESCRIPTION
**Pull request by ViktorThink**
Adding support for using Vicuna, Koala, Pythia-Chat-Base-7B, GPT-NeoXT-Chat-Base-20B for mining on netuid 1.

Closing my pervious pull requests [1266 ](https://github.com/opentensor/bittensor/pull/1266) and [1261](https://github.com/opentensor/bittensor/pull/1261) since this new pull request includes those models, and it was cleaner to have them all in the same branch.

Vicuna, Pythia and NeoXT has been tested that they work properly on netuid 1. I plan to test Koala today on netuid 1 if there are active validators.

### CODE REVIEW

**Comment by ViktorThink** [comment]
Now tested all four models today again, here are some random inputs/generations:

# Pythia
Message: [{'role': 'system', 'content': '\nYou are designed to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics.\n'}, {'role': 'user', 'content': '\nAsk me a random question about anything. Make the question very domain specific. Do not include the answer in the question.\n'}]
Generation: How many licks does it take to get to the tootsie roll center of a tootsie roll pop?

#NeoXT
Message: [{'role': 'system', 'content': '\nYou are designed to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics.\n'}, {'role': 'user', 'content': 'What is the difference between a class and an object in object-oriented programming?'}]
Generation: The fundamental difference between a class and an object in object-oriented programming is that a class defines the attributes and behaviors that all instances of a particular object have, while an object is a specific instance of a class that you can manipulate.

#Vicuna
Message: [{'role': 'system', 'content': '\nYou are designed to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics.\n'}, {'role': 'user', 'content': '\nAsk me a random question about anything. Make the question very domain specific. Do not include the answer in the question.\n'}]
Generation: What is the process for creating a new chemical compound in a laboratory setting?


Message: [{'role': 'system', 'content': '\nYou are designed to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics.\n'}, {'role': 'user', 'content': 'What is the process for creating a new chemical compound in a laboratory setting?'}]
Generation: The process for creating a new chemical compound in a laboratory setting typically involves several steps:

1. Designing the compound: This involves determining the structure and properties of the desired compound, as well as considering any potential challenges or obstacles that may arise during the synthesis process.
2. Gathering materials and equipment: The necessary materials and equipment will depend on the specific compound being synthesized, but may include reagents, solvents, glassware, and various pieces of equipment such as stirrers, heaters, and spectrophotometers.
3. Synthesizing the compound: This involves combining the appropriate starting materials in a series of steps, using various chemical reactions and techniques to build the desired structure.
4. Purifying the compound: After the compound has been synthesized, it may need to be purified to remove any impurities or unwanted byproducts. This may involve using techniques such as crystallization, chromatography, or recrystallization.
5. Characterizing the compound: Once the compound has been purified, it is important to determine its physical and chemical properties. This may involve using techniques

#Koala
Message: [{'role': 'system', 'content': '\nYou are designed to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics.\n'}, {'role': 'user', 'content': '\nAsk me a random question about anything. Make the question very domain specific. Do not include the answer in the question.\n'}]
Generation: What is the most effective way to treat a specific type of skin condition that is caused by a bacterial infection?



Message: [{'role': 'system', 'content': '\nYou are designed to assist with a wide range of tasks, from answering simple questions to providing in-depth explanations and discussions on a wide range of topics.\n'}, {'role': 'user', 'content': "What is the difference between a coroutine and a thread in Python's asyncio module?"}]
Generation: In Python's asyncio module, a coroutine is a lightweight form of concurrency that allows you to write asynchronous code without having to use a separate thread. A coroutine is a piece of code that can be executed asynchronously, meaning that it can run in the background while other code is being executed.

A thread, on the other hand, is a separate piece of code that runs on its own, independent of other threads. A thread can be used to perform tasks that require more processing power or memory than a coroutine can handle. For example, a thread can be used to perform a long-running calculation or to handle a large amount of data.

In summary, a coroutine is a lightweight form of concurrency that can be used to perform tasks that do not require a lot of processing power or memory, while a thread is a separate piece of code that can be used to perform tasks that require more processing power or memory.

**Comment by ViktorThink** [comment]
@joeylegere the PR is ready for merge, tested all four models today on the network and worked as expected.

I do plan to add these features:

- Add option to instead of using system prompts (messages from the role "system") use either the system prompt used during trainng, or a custom prompt.

- Add custom stop token for Pythia and NeoXT, to finish generation early.

File: neurons/text/prompting/miners/koala/README.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+# Converting Weights Into Model
+If you already have a converted checkpoint of the model, you can skip this step.
+
+Original documentation for creating the model from weights can be found [here](https://github.com/young-geng/EasyLM/blob/main/docs/koala.md)
+
+## Obtaining the Wegith Diff of Koala
+Due to the licence of the LLaMA model, the fine-tuned
+Koala model weights can not be directly released. Instead, the diff of weights, which can be used
+to recover the Koala model weights with the origina LLaMA model weights, is released. The diff
```
Comment (camfairchild) [inline neurons/text/prompting/miners/koala/README.md]:
```suggestion
to recover the Koala model weights with the original LLaMA model weights, is released. The diff
```
Typo

File: neurons/text/prompting/miners/neoxt/neuron.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
+# The MIT License (MIT)
+# Copyright © 2021 Yuma Rao
```
Comment (camfairchild) [inline neurons/text/prompting/miners/neoxt/neuron.py]:
```suggestion
# Copyright © 2023 Opentensor Foundation
```

File: neurons/text/prompting/miners/koala/neuron.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
+# The MIT License (MIT)
+# Copyright © 2021 Yuma Rao
```
Comment (camfairchild) [inline neurons/text/prompting/miners/koala/neuron.py]:
```suggestion
# Copyright © 2023 Opentensor Foundation
```

File: neurons/text/prompting/miners/vicuna/README.md [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
+This code is for running the Vicuna model through the BitTensor framework.
 
-## Vicuna Delta Miner
-Vicuna Chat Model miner for bittensor's prompting network.
-See: https://huggingface.co/lmsys/vicuna-13b-delta-v1.1 
+# Overview
 
+## Contents
+
+- [Installing Dependencies](#installing-Dependencies)
```
Comment (camfairchild) [inline neurons/text/prompting/miners/vicuna/README.md]:
```suggestion
- [Installing Dependencies](#installing-dependencies)
```

File: neurons/text/prompting/miners/vicuna/README.md [hunk 5]
```diff
@@ -26,15 +31,9 @@ You can add our delta to the original LLaMA weights to obtain the Vicuna weights
...
-
-### Install FastChat
-
-```bash
-# Install FastChat
-pip3 install fschat
-
-# Install the latest main branch of huggingface/transformers
-pip3 install git+https://github.com/huggingface/transformers
+Depending on which conversion script was used to create the Huggingface checkpoint of Llama, you might get an error that the tokenizer can not be found when loading the tokenizer. You can then replace all AutoTokenizers command with the correct tokenizer (in the exampel "LlamaTokenizer"), using this command:
```
Comment (camfairchild) [inline neurons/text/prompting/miners/vicuna/README.md]:
```suggestion
Depending on which conversion script was used to create the Huggingface checkpoint of Llama, you might get an error that the tokenizer can not be found when loading the tokenizer. You can then replace all AutoTokenizers command with the correct tokenizer (in the example "LlamaTokenizer"), using this command:
```

File: neurons/text/prompting/miners/vicuna/neuron.py [hunk 6]
```diff
@@ -29,51 +29,65 @@ def check_config( cls, config: 'bittensor.Config' ):
...
+
+        input_ids = self.tokenizer.encode(prompt, return_tensors="pt").to(self.config.vicuna.device)
+
+        output = self.model.generate(
+        input_ids,
+        max_length=input_ids.shape[1] + self.config.vicuna.max_new_tokens,
+        temperature=self.config.vicuna.temperature,
+        do_sample=self.config.vicuna.do_sample,
+        pad_token_id=self.tokenizer.eos_token_id,
+        )
```
Comment (camfairchild) [inline neurons/text/prompting/miners/vicuna/neuron.py]:
```suggestion
            output = self.model.generate(
            input_ids,
            max_length=input_ids.shape[1] + self.config.vicuna.max_new_tokens,
            temperature=self.config.vicuna.temperature,
            do_sample=self.config.vicuna.do_sample,
            pad_token_id=self.tokenizer.eos_token_id,
        )
```

**Review by camfairchild — COMMENTED** [review]
Seems good. Not sure about the license implications.

**Review by ifrit98 — APPROVED** [review]
Looks good.  Thanks for adding tweaks/suggestions @camfairchild

**Comment by ifrit98** [comment]
@ViktorThink Nice work! 

A few Qs: 
- Can you provide a complete example for running your Koala miner? Given the instructions and code provided, you cannot instantiate a Koala miner, even after converting the weights.
- Do you use a LLaMA tokenizer? If not, what tokenizer should be used with this model? (unclear)


Command ran:
```
python neurons/text/prompting/miners/koala/neuron.py --koala.model_name /home/jason/sandbox/PR/llama/llama_weights/koala/7B/model --netuid 1 --logging.debug --wallet.hotkey miner4
```

Error message (indicating there is no tokenizer found, which makes sense because there is not one mentioned in the readme at all and is not discussed how to move from the final Koala weights to prepare the miner.)
```
HFValidationError: Repo id must be in the form 'repo_name' or 'namespace/repo_name': '/home/jason/sandbox/PR/llama/llama_weights/koala/7B/model'. Use `repo_type` argument if needed.
```

The error is on line 45:
```self.tokenizer = AutoTokenizer.from_pretrained( self.config.koala.model_name, use_fast=False )```

[https://github.com/opentensor/bittensor/blob/b179b334afc16ed7a8f654fcc1b350abcf559ef8/neurons/text/prompting/miners/koala/neuron.py#L45](url)

**Comment by ViktorThink** [comment]
@ifrit98 Ah, nice spotting. Yes, it is the same tokenizer as llama. I'll update the documentation later today to show exactly how to do it.

**Comment by ViktorThink** [comment]
@ifrit98  I had missed to add the last step of the conversion process before starting the miner.

> ## Converting the Koala Weights to HuggingFace Transformers
> To run the model on the miner, it has to be converted to Huggingface Transformers format. To do so, use
> the following command:
> 
> ``` shell
> python -m EasyLM.models.llama.convert_easylm_to_hf \
>     --load_checkpoint='params::path/to/koala/checkpoint' \
>     --tokenizer_path='path/to/llama/tokenizer' \
>     --model_size='13b' \  # '7b', '13b', '30b' or '65b'
>     --output_dir='path/to/output/huggingface/koala/checkpoint'
> ```
> 
> # Starting Miner
> ```
> python3 neurons/text/prompting/miners/koala/neuron.py --koala.model_name model_path_or_huggingface_identifier
> ```

EasyLM has not released any converted version of the models, only delta, but I saw on Huggingface there are already converted checkpoints that could be loaded directly.

I don't actually have the Llama weights in torch format, so I will look into adding a second option for conversion for those who have Llama in HF format but not in torch.

**Comment by ifrit98** [comment]
Converted and miner is up and running.  Nice work! Closing.
